packages/local_auth/local_auth_android/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java

```diff
@@ -147,79 +147,104 @@ private void authenticate(MethodCall call, final Result result) {
-              lifecycle, (FragmentActivity) activity, call, completionHandler, false);
-      authHelper.authenticate();
-      return;
-    }
+    ArrayList<String> authenticationMethods = getEnrolledBiometrics();
 
-    // API 29 and above
-    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
-      authHelper =
-          new AuthenticationHelper(
-              lifecycle, (FragmentActivity) activity, call, completionHandler, true);
-      authHelper.authenticate();
+    // Handle case where any biometrics are available
+    if (canAuthenticateWithBiometrics()) {
+      if (!isBiometricOnly && canAuthenticateWithDeviceCredential()) {
+        authenticationMethods.add("deviceCredential");
+      }
+      sendAuthenticationRequest(call, completionHandler, authenticationMethods);
       return;
     }
 
-    // API 23 - 28 with fingerprint
-    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && fingerprintManager != null) {
-      if (fingerprintManager.hasEnrolledFingerprints()) {
-        authHelper =
-            new AuthenticationHelper(
-                lifecycle, (FragmentActivity) activity, call, completionHandler, false);
-        authHelper.authenticate();
+    // Handle case where only device credential is available
+    if (!isBiometricOnly && canAuthenticateWithDeviceCredential()) {
+      if (Build.VERSION.SDK_INT >= 30) {
+        sendAuthenticationRequest(
+            call, completionHandler, new ArrayList<String>(Arrays.asList("deviceCredential")));
+        return;
+      } else if (Build.VERSION.SDK_INT >= 23
+          && keyguardManager != null
+          && keyguardManager.isDeviceSecure()) {
+        String title = call.argument("signInTitle");
+        String reason = call.argument("localizedReason");
+        Intent authIntent = keyguardManager.createConfirmDeviceCredentialIntent(title, reason);
+
+        // Save result for async response
+        lockRequestResult = result;
+        activity.startActivityForResult(authIntent, LOCK_REQUEST_CODE);
         return;
       }
     }
 
-    // API 23 or higher with device credentials
-    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M
-        && keyguardManager != null
-        && keyguardManager.isDeviceSecure()) {
-      String title = call.argument("signInTitle");
-      String reason = call.argument("localizedReason");
-      Intent authIntent = keyguardManager.createConfirmDeviceCredentialIntent(title, reason);
-
-      // save result for async response
-      lockRequestResult = result;
-      activity.startActivityForResult(authIntent, LOCK_REQUEST_CODE);
-      return;
+    // Unable to authenticate
+    int errorCode =
+        isBiometricOnly
+            ? biometricManager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_WEAK)
+            : biometricManager.canAuthenticate(
+                BiometricManager.Authenticators.BIOMETRIC_WEAK
+                    | BiometricManager.Authenticators.DEVICE_CREDENTIAL);
+
+    switch (errorCode) {
```

Comment:
I added more error handling below based on the error messages of the new API. I changed some of the error messages, so let me know if those are magic strings and need to be changed back or added elsewhere.

Reply:
I don't think we should consider the message itself to be API-stable, just the code. (Theoretically someone could be relying on it of course, but they shouldn't be.)